### PR TITLE
kvserver: add sstable compression algorithm metrics

### LINKED ISSUE #123952: storage: add a timeseries metric for SST count by compression algorithm
Expose a timeseries metric that will allow us to slice the SSTable count by compression algorithm (snappy vs zstd).

Jira issue: CRDB-38624

### PR DESCRIPTION
**Pull request by nicktrav**
Add metrics tracking the count of SSTables using various compression algorithms:

- `storage.sstable.compression.snappy.count`
- `storage.sstable.compression.zstd.count`
- `storage.sstable.compression.unknown.count`
- `storage.sstable.compression.none.count`

Closes #123952.

Release note: None.

### CODE REVIEW

**Comment by nicktrav** [comment]
- Started up with the default (`snappy`)
- Switched to zstd: `set cluster setting storage.sstable.compression_algorithm = 'zstd';`
- Ran a manual compaction: `SELECT crdb_internal.compact_engine_span(1, 1, decode('00', 'hex'), decode('FFFFFFFF', 'hex'));`

Screenshot 2024-06-27 at 7 58 27 AM

**Comment by RaduBerinde** [comment]
Do we care about the count or do we care about the total sizes more? This will be more relevant especially if we add heuristics (e.g. use zstd only for L6).

File: pkg/kv/kvserver/metrics.go [hunk 1]
```diff
@@ -963,6 +963,32 @@ bytes preserved during flushes and compactions over the lifetime of the process.
 		Measurement: "Bytes",
 		Unit:        metric.Unit_BYTES,
 	}
+	metaSSTableCompressionSnappy = metric.Metadata{
+		Name: "storage.sstable.compression.snappy.count",
+		Help: "Count of SSTables that have compressed with the snappy compression " +
```
Comment (itsbilal) [inline pkg/kv/kvserver/metrics.go:968]: nit: "have been"

**Review by itsbilal — APPROVED** [review]
LGTM

Personally I'm okay with a count as opposed to bytes (plus we'd need to update the plumbing in Pebble to reflect bytes). If we do go towards a level-based compression heuristic, we can update the metric in Pebble.

**Review by RaduBerinde — APPROVED** [review]
:lgtm:

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/126330#-:-O0VWyBV7X3kOZFVBC9g:bnfp4nl)* status: <span>:shipit:</span> complete! 1 of 0 LGTMs obtained (waiting on @jbowens and @nicktrav)

___
*[`pkg/kv/kvserver/metrics.go` line 981 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/126330#-O0VWoPe3DEV3r8iwnNl:-O0VWoPe3DEV3r8iwnNm:b-4jxjs) ([raw file](https://github.com/cockroachdb/cockroach/blob/10b4e48a8b0177c172903df73c63450561c3895a/pkg/kv/kvserver/metrics.go#L981)):*
> ```Go
> 	}
> 	metaSSTableCompressionUnknown = metric.Metadata{
> 		Name:        "storage.sstable.compression.unknown.count",
> ```

Do we really need to export this metric? We wouldn't be able to use these tables, it would be a bug no?

**Comment by nicktrav** [comment]
TFTRs!

bors r=itsbilal

**Comment by nicktrav** [comment]
> We wouldn't be able to use these tables, it would be a bug no?

I included them for completeness (i.e. sum(per-compression alg totals) == total sstables metric). If those series were non-zero, then yes, I think it would indicate something to dig into.

**Comment by nicktrav** [comment]
bors cancel

**Comment by nicktrav** [comment]
bors r=itsbilal,RaduBerinde

**Pull request merged**